test/command-fsck.test

```diff
@@ -0,0 +1,289 @@
+  cat >"$foo_dir/ec_pubkey" <<EOF
+-----BEGIN ED25519 PUBLIC KEY-----
+$(awk '/^Ed25519PublicKey/ { printf $NF }' "$foo_host")
+-----END ED25519 PUBLIC KEY-----
+EOF
+  sed_cmd '/Ed25519PublicKey/d' "$foo_host"
+  echo "Ed25519PublicKeyFile = $(normalize_path "$foo_dir/ec_pubkey")" >>"$foo_dir/$conf"
+  fail_on_msg 'no (usable) public Ed25519' tinc foo fsck
+}
+
+test_rsa_public_key_file_var() {
+  conf=$1
+  fsck_test "RSA public key in PublicKeyFile in $conf must work"
+  extract_pem_key_from_config "$foo_host" >"$foo_dir/rsa_pubkey"
+  rm_pem_key_from_config "$foo_host"
+  echo "PublicKeyFile = $(normalize_path "$foo_dir/rsa_pubkey")" >>"$foo_dir/$conf"
+  fail_on_msg 'error reading RSA public key' tinc foo fsck
+}
+
+fsck_test 'Newly created configuration should pass'
+tinc foo fsck
+
+fsck_test 'Must fail on missing tinc.conf'
+rm -f "$foo_conf"
+must_fail_with_msg 'no tinc configuration found' tinc foo fsck
+
+if run_access_checks; then
+  fsck_test 'Must fail on inaccessible tinc.conf'
+  chmod 000 "$foo_dir"
+  must_fail_with_msg 'not running tinc as root' tinc foo fsck
+fi
+
+if ! is_windows; then
+  fsck_test 'Non-executable tinc-up MUST be fixed by tinc --force'
+  chmod a-x "$foo_tinc_up"
+  expect_msg 'cannot read and execute' tinc foo --force fsck
+  test -x "$foo_tinc_up"
+
+  fsck_test 'Non-executable tinc-up MUST NOT be fixed by tinc without --force'
+  chmod a-x "$foo_tinc_up"
+  expect_msg 'cannot read and execute' tinc foo fsck
+  must_fail test -x "$foo_tinc_up"
+fi
+
+fsck_test 'Unknown -up script warning'
+touch "$foo_dir/fake-up"
+expect_msg 'unknown script' tinc foo fsck
+
+fsck_test 'Unknown -down script warning'
+touch "$foo_dir/fake-down"
+expect_msg 'unknown script' tinc foo fsck
+
+if ! is_windows; then
+  fsck_test 'Non-executable foo-up MUST be fixed by tinc --force'
+  touch "$foo_host_up"
+  chmod a-x "$foo_host_up"
+  expect_msg 'cannot read and execute' tinc foo --force fsck
+  test -x "$foo_tinc_up"
+
+  fsck_test 'Non-executable bar-up MUST NOT be fixed by tinc'
+  touch "$foo_dir/hosts/bar-up"
+  chmod a-x "$foo_dir/hosts/bar-up"
+  expect_msg 'cannot read and execute' tinc foo fsck
+  must_fail test -x "$foo_dir/bar-up"
+fi
+
+if run_access_checks; then
+  fsck_test 'Inaccessible hosts/foo must fail'
+  chmod 000 "$foo_host"
+  must_fail_with_msg 'cannot open config file' tinc foo fsck
+fi
+
+fsck_test 'Must fail when all private keys are missing'
+rm -f "$foo_ec_priv" "$foo_rsa_priv"
+if with_legacy; then
+  must_fail_with_msg 'neither RSA or Ed25519 private key' tinc foo fsck
+else
+  must_fail_with_msg 'no Ed25519 private key' tinc foo fsck
+fi
+
+if with_legacy; then
+  test_private_keys rsa_key.priv
+
+  if ! is_windows; then
+    fsck_test 'Must warn about unsafe permissions on tinc.conf with PrivateKey'
+    rm -f "$foo_rsa_priv"
+    echo "PrivateKey = $rsa_d" >>"$foo_conf"
+    echo "PublicKey = $rsa_n" >>"$foo_host"
+    expect_msg 'unsafe file permissions' tinc foo fsck
```

Comment:
This is failing if I run it locally, since my umask is set to 077 and the files are created with safe permissions. You should explicitly chmod these files to something unsafe.

Reply:
Good catch, thanks. Fixed. In addition to umask, I'll think about other non-standard configurations to add to CI.